fix: NetworkAnimator Does not unsubscribe from  OnClientConnectedCallback when despawned.

NoelStephensUnity opened the pull request:
This fixes the issue where NetworkAnimator was not unsubscribing from the OnClientConnectedCallback when despawned.
This also fixed some minor testproject related issues 
[MTT-4080](https://jira.unity3d.com/browse/MTT-4080)

## Changelog
- Fixed: issue where NetworkAnimator was not unsubscribing from OnClientConnectedCallback when despawned. 

## Testing and Documentation
- Includes integration test to verify this fix.

NoelStephensUnity committed:
fix

This resolves the issue with NetworkAnimator not de-registering itself from OnClientConnected.
This resolves some issues with the manual NetworkAnimator test.

NoelStephensUnity committed:
update

MTT-4080
(this branch includes fixes for the OnClientConnectedCallback issue)
Renamed NetworkAnimatorEnhancement scene to NetworkAnimatorServerOwnerTest.
Adding the manual NetworkAnimatorServerOwnerTest scene to the scenes in build list.

NoelStephensUnity committed:
update

MTT-4080
Updating the changelog.

NoelStephensUnity committed:
style

fixing whitespace issue.

NoelStephensUnity committed:
Merge branch 'develop' into fix/network-animator-unsubscribe-always

NoelStephensUnity committed:
test

Adding the integration test that verifies NetworkAnimator properly removes its subscription to the NetworkManager.OnClientConnectedCallback when it is despawned.

jeffreyrainy commented on testproject/Assets/Tests/Manual/NetworkAnimatorTests/AnimatedCubeController.cs:
```diff
@@ -13,32 +13,34 @@ public class AnimatedCubeController : NetworkBehaviour
+            }
             m_Animator = GetComponent<Animator>();
-            m_NetworkAnimator = GetComponent<NetworkAnimator>();
+            m_NetworkAnimator = GetComponent<OwnerNetworkAnimator>();
```
The fact that the order was wrong, combined with the low readability (assigning different component to the same member variable?!) points to the need for a comment. :-)

Maybe something above saying "We want to find either the OwnerNetworkanimator or the regular NetworkAnimator, with preference for the OwnerNetworkanimator".

Just a suggestion.

jeffreyrainy commented on testproject/Assets/Tests/Manual/NetworkAnimatorTests/AnimatedCubeController.cs:
```diff
@@ -13,32 +13,34 @@ public class AnimatedCubeController : NetworkBehaviour
                 if (m_NetworkAnimator != null)
                 {
-                    m_IsServerAuthoritative = false;
+                    m_IsServerAuthoritative = true;
```
Can we have `m_IsServerAuthoritative = (m_NetworkAnimator != null);` outside the if, instead? As-is, it gives the impression that if `m_NetworkAnimator` is not null, `m_IsServerAuthoritative` keeps it previous value, which forces one to then wonder "is it a default false, or does it persist previous value?"

jeffreyrainy commented on testproject/Assets/Tests/Manual/NetworkAnimatorTests/AnimationUserInput.cs:
```diff
@@ -26,15 +26,21 @@ private void LateUpdate()
                 foreach (var controller in m_Controllers)
                 {
-                    controller.ToggleRotateAnimation();
+                    if ((controller.IsServerAuthoritative && (IsServer || IsOwner)) || (!controller.IsServerAuthoritative && IsOwner))
```
Just checking. when `IsOwner` and `IsServerAuthoritative` we still want this ?

NoelStephensUnity replied:
Actually, all of that is not needed when checking for user input. The conditional requirement to check for user input was culled down to:
- IsSpawned has to be true
- IsOwner has to be true
If neither of the above, then it exist early.  Otherwise, it checks for any user-owner input.
_(another good catch... my brain is still switching from "documentation" mode to "coding mode"...lol)_

NoelStephensUnity committed:
update

Fixing how AnimatedCubeController determines the type of NetworkAnimator and whether it is server or owner authoritative.

NoelStephensUnity committed:
update

Simplifying the conditional check requirements down to IsSpawned and IsOwner for the AnimationUserInput component since the only input checked should always be by the owner.

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -268,10 +268,7 @@ private void Cleanup()
-            {
-                NetworkManager.OnClientConnectedCallback -= OnClientConnectedCallback;
-            }
+            NetworkManager.OnClientConnectedCallback -= OnClientConnectedCallback;
```
can we have this change only in isolation or does changes in testproject are also needed?

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -5,6 +5,15 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 Additional documentation and release notes are available at [Multiplayer Documentation](https://docs-multiplayer.unity3d.com).
+## [Unreleased]
```
there's already an unreleased tag below :)

NoelStephensUnity replied:
That must have been a merge issue...when I added that there was none.

NoelStephensUnity committed:
update

fixing merge issue with changelog.
